torch/csrc/autograd/engine.cpp

```diff
@@ -201,8 +216,22 @@ auto ReadyQueue::pop() -> FunctionTask {
 
 Engine::Engine() = default;
 
-// This Engine's ReadyQueues and their corresponding threads are leaked here
-Engine::~Engine() = default;
+// Send shutdown tasks to all ReadyQueues if no backward tasks are running
+// Even though readyQueue should be empty, shutdown tasks have the highest
+// priority
+Engine::~Engine() {
+  bool noBackward = true;
+  for (auto& queue: ready_queues_) {
+    std::lock_guard<std::mutex> lock(queue->mutex_);
```

Comment:
Taking out a lock on a mutex in a destructor makes me nervous for a few reasons:
1. Taking out a lock could throw an exception. Throwing exceptions in a destructor is bad juju, because you won't actually get an exception; you'll just forcibly terminate the program.
2. It could lead to a deadlock at program shutdown (no one's idea of a happy time). I think in this case it's not possible, because the scope we acquire `mutex_` is not very large and we don't ever take out other locks.

It seems to me that all we need is a "best effort" test to see if we're running a backwards or not. Perhaps a single, global atomic counter on `Engine` itself would be good enough? Increment it when you start a backwards, decrement it when you finish (probably want to do this with RAII to make sure exceptions are handled correctly). Then look at that counter to decide if a backwards is running or not.

Reply:
Why do we take the lock at every loop iteration instead of acquiring it once before the loop?

Reply:
Also, are we sure that the destruction is guaranteed not to run on any of the worker threads? Like, what happens if one of them calls `os.exit`? Wouldn't that lead to case 2.?

Reply:
The lock has to be taken out on each iteration of the loop, because we don't have a global lock, it's a per ready queue lock, and so there isn't "one" lock to take out.

> Also, are we sure that the destruction is guaranteed not to run on any of the worker threads? Like, what happens if one of them calls os.exit?

I think you're right. If a worker thread calls `std::exit`, that thread will handle destructing static objects, and we'd have to detect that case in the destructor in that case.

This is all a bit moot right now, because this PR got disabled again when we landed the fix for infinite recursion. But if we do revive this patch it would be good to get this part right too.